perception/map_based_prediction/src/path_generator.cpp

```diff
@@ -0,0 +1,346 @@
+{
+  if (ref_paths.size() < 2) {
+    const PredictedPath empty_path;
+    return empty_path;
+  }
+
+  return generatePolynomialPath(origin_header, object, ref_paths);
+}
+
+PredictedPath PathGenerator::generateStraightPath(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object) const
+{
+  const auto & object_pose = object.kinematics.pose_with_covariance.pose;
+  const auto & object_twist = object.kinematics.twist_with_covariance.twist;
+
+  PredictedPath path;
+  path.time_step = rclcpp::Duration::from_seconds(sampling_time_interval_);
+  const double ep = 0.001;
+  for (double dt = 0.0; dt < time_horizon_ + ep; dt += sampling_time_interval_) {
+    geometry_msgs::msg::Pose object_frame_pose;
+    geometry_msgs::msg::Pose world_frame_pose;
+    object_frame_pose.position.x = object_twist.linear.x * dt;
+    object_frame_pose.position.y = object_twist.linear.y * dt;
+    tf2::Quaternion quat;
+    quat.setRPY(0.0, 0.0, 0.0);
+    object_frame_pose.orientation = tf2::toMsg(quat);
+    tf2::Transform tf_object2future;
+    tf2::Transform tf_world2object;
+    tf2::Transform tf_world2future;
+
+    tf2::fromMsg(object_pose, tf_world2object);
+    tf2::fromMsg(object_frame_pose, tf_object2future);
+    tf_world2future = tf_world2object * tf_object2future;
+    tf2::toMsg(tf_world2future, world_frame_pose);
+
+    path.path.push_back(world_frame_pose);
+  }
+
+  return path;
+}
+
+PredictedPath PathGenerator::generatePolynomialPath(
+  const std_msgs::msg::Header & origin_header, const TrackedObject & object,
+  const PosePath & ref_path)
+{
+  // Get current Frenet Point
+  const double ref_path_len = tier4_autoware_utils::calcArcLength(ref_path);
+  const auto current_point = getFrenetPoint(origin_header, object, ref_path);
+
+  // Step1. Set Target Frenet Point
+  // Note that we do not set position s,
+  // since we don't know the target longitudinal position
+  const double T = time_horizon_;
+  FrenetPoint terminal_point;
+  terminal_point.header = current_point.header;
+  terminal_point.header.stamp =
+    rclcpp::Time(current_point.header.stamp) + rclcpp::Duration::from_seconds(T);
+  terminal_point.s_vel = current_point.s_vel;
+  terminal_point.s_acc = 0.0;
+  terminal_point.d = 0.0;
+  terminal_point.d_vel = 0.0;
+  terminal_point.d_acc = 0.0;
+
+  // Step2. Generate Predicted Path on a Frenet coordinate
+  const auto frenet_predicted_path =
+    generateFrenetPath(current_point, terminal_point, ref_path_len);
+
+  // Step3. Interpolate Reference Path for converting predicted path coordinate
+  const auto interpolated_ref_path = interpolateReferencePath(ref_path, frenet_predicted_path);
+
+  if (frenet_predicted_path.size() < 2 || interpolated_ref_path.size() < 2) {
+    const PredictedPath empty_path;
+    return empty_path;
+  }
+
+  // Step4. Convert predicted trajectory from Frenet to Cartesian coordinate
+  return convertPredictedPath(object, frenet_predicted_path, interpolated_ref_path);
```

Comment:
In my understanding,
in step 2, s and d are calculated,
in step 3, s is applied to ref_path,
in step4, d is applied to ref_path (and yaw is calculated as well).

If so, the function names are very misleading to me.
It may be better for the names like this? What do you think.
convertReferencePathLongitudinally in step3
convertReferencePathLaterally in step4 (yaw calculation can be here, or you can make another function like alignYaw as a step5)

Reply:
In step3, this function just interpolates its arc length along with the frenet path generated in step2. So we basically do not convert any trajectory or path here. 

In step4, this function generates (x, y) coordinate trajectory based on frenet trajectory (s, d). Therefore it converts longitudianl and lateral coordinates at the same time.

Reply:
I don't understand well, but I think it's ok :+1: